### PR TITLE
[NT-1892] UI for comment card

### PR DESCRIPTION
**Pull request by moderateepheezy**
# 📲 What

Comment UI Card

# 🤔 Why

Part of the new Design for Threading Comment.

# 👀 See

Comment

# ✅ Acceptance criteria

1. Launch the app
2. Tap on any project card
3. Tap on comment section to show the new comments screen
4. Layout should align with https://www.figma.com/file/zyeDBH6vk2DMmt39S1yBEx/Native-Comment-Threading?node-id=2%3A3211
5. On Device setting change accessibility large text as desired, Layout display shouldn't be distorted.

# ⏰ TODO

- [x] Add indicator icon to comment removed text.
- [x] Create a reusable Style for all CommentCells
- [x] Remove DemoComment model and replace with Comment model
- [x] Update Unit Test
- [x] Fix Snapshot test on CommentViewController

### CODE REVIEW

**Comment by nativeksr** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Review by msadoon — COMMENTED** [review]
Yea this is amazing,
I just read it to get a feel for how I would update my [PR](https://github.com/kickstarter/ios-oss/pull/1477). 

Just general feedback,
I noticed one thing on super large accessibility text (reply text hidden).


Whenever we decide to merge this guy, I'll pull it into my PR to complete the pagination ticket 👍🏿 cc @justinswart

File: Library/ViewModels/CommentCellViewModelTests.swift
Comment (moderateepheezy) [reply Library/ViewModels/CommentCellViewModelTests.swift:27]: I used the function name because it seems like standard in the codebase.

File: Kickstarter-iOS/Views/Cells/CommentPostFailedCell.swift
Comment (moderateepheezy) [reply Kickstarter-iOS/Views/Cells/CommentPostFailedCell.swift:76]: So `override` function can't be `private`

File: Kickstarter-iOS/Views/Cells/CommentRemovedCell.swift
Comment (moderateepheezy) [reply Kickstarter-iOS/Views/Cells/CommentRemovedCell.swift]: Will keep note of this.

File: Kickstarter-iOS/Views/Controllers/CommentsViewController.swift
Comment (moderateepheezy) [reply Kickstarter-iOS/Views/Controllers/CommentsViewController.swift:25]: Yes, but that will go away when we retrieve comments from the `CommentViewModel`

**Comment by moderateepheezy** [comment]
> Yea this is amazing,
> I just read it to get a feel for how I would update my [PR](https://github.com/kickstarter/ios-oss/pull/1477).
> 
> Just general feedback,
> I noticed one thing on super large accessibility text (reply text hidden).
> Simulator Screen Shot - iPhone 8 - 2021-05-19 at 14 33 51
> 
> Whenever we decide to merge this guy, I'll pull it into my PR to complete the pagination ticket 👍🏿 cc @justinswart

@msadoon I'm not able to fix large accessibility text. Do you know of a way to fix this for a UIButton.

File: Kickstarter-iOS/DataSources/CommentsDataSourceTests.swift
Comment (justinswart) [reply Kickstarter-iOS/DataSources/CommentsDataSourceTests.swift:14]: We do have `Comment.template` in the codebase already 🤔

File: Kickstarter-iOS/Views/Cells/CommentCell.swift
Comment (justinswart) [inline Kickstarter-iOS/Views/Cells/CommentCell.swift:9]: `private` is fine, best to only use `fileprivate` if we need to.

File: Kickstarter-iOS/Views/Cells/CommentCell.swift
Comment (justinswart) [inline Kickstarter-iOS/Views/Cells/CommentCell.swift:63]: Not sure why we need `DemoComment`, the `Comment` model does exist in the codebase. If you need to access the template, just import KsApi with `@testable`. Either way let's just be sure not to merge `DemoComment` 👍

File: Kickstarter-iOS/Views/Cells/CommentCell.swift
Comment (justinswart) [inline Kickstarter-iOS/Views/Cells/CommentCell.swift:83]: Not sure if we need any of these constants, we should be able to float these elements into place?

File: Kickstarter-iOS/Views/Cells/CommentPostFailedCell.swift
Comment (justinswart) [inline Kickstarter-iOS/Views/Cells/CommentPostFailedCell.swift:96]: `7.17` is very specific? 😄

File: Kickstarter-iOS/Views/Cells/CommentRemovedCell.swift
Comment (justinswart) [inline Kickstarter-iOS/Views/Cells/CommentRemovedCell.swift:85]: Does the image not have an intrinsic size?

File: Kickstarter-iOS/Views/Cells/CommentRemovedCell.swift
Comment (justinswart) [inline Kickstarter-iOS/Views/Cells/CommentRemovedCell.swift:120]: Will this be made into a link in a different PR?

File: Kickstarter-iOS/Views/CommentCellHeaderStackView.swift
Comment (justinswart) [inline Kickstarter-iOS/Views/CommentCellHeaderStackView.swift:78]: Typo on the function name.

Also, usually we would output something like this from the VM and write tests for it. I guess if we are confident this is covered by snapshot tests we could forgo it here 🤔 @singhhari ?

File: Kickstarter-iOS/Views/Controllers/CommentsViewController.swift
Comment (justinswart) [inline Kickstarter-iOS/Views/Controllers/CommentsViewController.swift:34]: Do we want this hidden?

File: Library/ViewModels/CommentCellViewModel.swift
Comment (justinswart) [inline Library/ViewModels/CommentCellViewModel.swift:40]: If you want you can use key paths here:
```swift
self.body = comment.map(\.body)
```

**Review by justinswart — COMMENTED** [review]
Looking good! Code is nice and neat 👍 

Couple of comments (😄) and questions

**Review by singhhari — COMMENTED** [review]
A couple of observations:
- The text view is selectable, currently, so if I click on it the cursor shows.
- Raising the font size larger through the Accessibility Inspector results in the title label of the cell being illegible. I think those constraints might need to be updated.

File: Kickstarter-iOS/Views/Cells/CommentCell.swift
Comment (singhhari) [inline Kickstarter-iOS/Views/Cells/CommentCell.swift:16]: alphabetize properties

File: Kickstarter-iOS/Views/Cells/CommentRemovedCell.swift
Comment (singhhari) [reply Kickstarter-iOS/Views/Cells/CommentRemovedCell.swift:120]: Typo in the name of the variable here

File: Library/ViewModels/CommentCellViewModel.swift
Comment (singhhari) [inline Library/ViewModels/CommentCellViewModel.swift:18]: alphabetize

File: Kickstarter-iOS/Views/Cells/CommentPostFailedCell.swift
Comment (moderateepheezy) [reply Kickstarter-iOS/Views/Cells/CommentPostFailedCell.swift:96]: Haaa, the figma design has that specific margin but I have changed it to `Styles.grid(6)` since it's much more closer to `7.17`

File: Kickstarter-iOS/Views/Cells/CommentRemovedCell.swift
Comment (moderateepheezy) [reply Kickstarter-iOS/Views/Cells/CommentRemovedCell.swift:85]: So because of how the design is on Figma, the width and height would need to be fixed, so the image can float at the top. Unless there's a better way to approach this.

File: Kickstarter-iOS/Views/Cells/CommentRemovedCell.swift
Comment (moderateepheezy) [reply Kickstarter-iOS/Views/Cells/CommentRemovedCell.swift:120]: I was thinking it should go in a different PR, but seeing that it won't take much time to fix, I'll be adding it to this PR

File: Kickstarter-iOS/Views/CommentCellHeaderStackView.swift
Comment (moderateepheezy) [reply Kickstarter-iOS/Views/CommentCellHeaderStackView.swift:78]: Yeah, I thought About this too, but I was thinking how to test this in the `vm`, and since snapshot test should cover it, I don't need it in the vm. 
But then we need to be confident that snapshot tests will cover it.

File: Kickstarter-iOS/Views/Controllers/CommentsViewController.swift
Comment (moderateepheezy) [reply Kickstarter-iOS/Views/Controllers/CommentsViewController.swift:34]: Yes otherwise we won't see the NavBar.

Comment (justinswart) [reply Kickstarter-iOS/Views/Controllers/CommentsViewController.swift:34]: oh sorry missed that it's being set to `false`.... ok, will need to just test that this doesn't affect other areas.

File: Kickstarter-iOS/DataSources/CommentsDataSourceTests.swift
Comment (moderateepheezy) [reply Kickstarter-iOS/DataSources/CommentsDataSourceTests.swift]: Removed `DemoComment`, and also created some extra templates to mock the UI.

File: Kickstarter-iOS/Views/CommentCellHeaderStackView.swift
Comment (moderateepheezy) [reply Kickstarter-iOS/Views/CommentCellHeaderStackView.swift]: I have moved this to the VM and provided tests for it.

**Comment by msadoon** [comment]
@moderateepheezy All my comments are resolved, so thank you! Can you quickly check the padding on the left/right of the `CommentCell` against the tableview leading/trailing? Jonathon noticed the Figma has it at 18 and in this snapshot it looks like 36. It does look a little larger than in the design.

File: Kickstarter-iOS/Views/Cells/CommentRemovedCell.swift
Comment (moderateepheezy) [reply Kickstarter-iOS/Views/Cells/CommentRemovedCell.swift]: So I think this would need to be in a different ticket, there's currently a ticket raised for localization which I believe @singhhari is assigned to.
The project has a template to work with links from text and it seems to be tied with localized string.

**Comment by moderateepheezy** [comment]
@msadoon  updated the PR, you can let me know if the margin is resolved.

**Comment by msadoon** [comment]
> @msadoon updated the PR, you can let me know if the margin is resolved.

Looks good now! cc. @jgsamudio 

![]()

**Review by singhhari — COMMENTED** [review]
Rotating the screen results in the badges distorting in layout and color

File: Kickstarter-iOS/Views/Cells/CommentPostFailedCell.swift
Comment (singhhari) [inline Kickstarter-iOS/Views/Cells/CommentPostFailedCell.swift:21]: maybe just `retryButton` ?

File: Kickstarter-iOS/Views/CommentCellHeaderStackView.swift
Comment (singhhari) [inline Kickstarter-iOS/Views/CommentCellHeaderStackView.swift:11]: alphabetize

File: KsApi/models/Comment.swift
Comment (singhhari) [inline KsApi/models/Comment.swift:22]: I think this is fine but should run it by Paulo so Product is aware of this convention

File: KsApi/models/graphql/adapters/Comment+GraphComment.swift
Comment (singhhari) [inline KsApi/models/graphql/adapters/Comment+GraphComment.swift:22]: I'm a little unclear about the `FIXME` comments I'm seeing related to the mapping of Comment to GraphComment. Is there a reason we are defaulting values instead of making them optionals set to nil?

File: Library/PaddingLabel.swift
Comment (singhhari) [inline Library/PaddingLabel.swift:5]: Could you tell me a bit about why we are using this custom UILabel instead of applying a style to a UILabel?

File: Library/ViewModels/CommentCellViewModel.swift
Comment (singhhari) [inline Library/ViewModels/CommentCellViewModel.swift:12]: typo `Alignment`

File: Library/ViewModels/CommentCellViewModelTests.swift
Comment (singhhari) [inline Library/ViewModels/CommentCellViewModelTests.swift:46]: typo `emitted`

File: Library/ViewModels/CommentCellViewModelTests.swift
Comment (singhhari) [inline Library/ViewModels/CommentCellViewModelTests.swift:85]: typo `emitted`

File: Library/ViewModels/CommentCellViewModelTests.swift
Comment (singhhari) [reply Library/ViewModels/CommentCellViewModelTests.swift:46]: there's a few instances of this. prob best to do a find and replace.

File: KsApi/models/Comment.swift
Comment (singhhari) [reply KsApi/models/Comment.swift:22]: Resolving since we are continuing this on the following thread:
https://kickstarter.slack.com/archives/CG55BBBBL/p1621968220061600

File: Library/PaddingLabel.swift
Comment (moderateepheezy) [reply Library/PaddingLabel.swift:5]: Oh yeah. Intrinsic content size of a label doesn't have padding, if you add a background colour it won't look like the design,  so to make the AuthorBadge look like the design with the padding I'll need to subclass UILabel.

**Review by msadoon — APPROVED** [review]

File: Kickstarter-iOS/Views/Cells/CommentRemovedCell.swift
Comment (singhhari) [reply Kickstarter-iOS/Views/Cells/CommentRemovedCell.swift]: I'll update the localization ticket to include this here but leave a TODO/FIXME so I can search for it

File: Kickstarter-iOS/Views/Cells/CommentCell.swift
Comment (justinswart) [inline Kickstarter-iOS/Views/Cells/CommentCell.swift:96]: The key and default value should match with the key having underscores. so in this case:

```swift
localizedString(key: "Reply", defaultValue: "Reply")
```

That way when we add this to the strings for translation it'll become `Strings.Reply()`.

File: Kickstarter-iOS/Views/Cells/CommentRemovedCell.swift
Comment (justinswart) [reply Kickstarter-iOS/Views/Cells/CommentRemovedCell.swift:85]: Instead of width constraints here can we use:

```swift
self.infoImageView.setContentHuggingPriority(.required, for: .horizontal)
```

File: Kickstarter-iOS/Views/CommentCellHeaderStackView.swift
Comment (justinswart) [inline Kickstarter-iOS/Views/CommentCellHeaderStackView.swift:12]: This is happening because the style is being emitted from the view model. We either need to call an input on the view model when rotation occurs so that it emits again or keep the styles in the cell and test them with snapshots. I think it's best to just rely on our snapshots here. Let's fix this before merging this PR because it may cause confusion later on.

It seems that a user can have multiple badges, this wasn't really catered for in the designs so we should just confirm what the priority of badges are (creator -> superbacker) etc.

File: Library/Styles/CommentCellStyles.swift
Comment (justinswart) [inline Library/Styles/CommentCellStyles.swift:53]: Key and value to match here.

File: Library/Styles/CommentCellStyles.swift
Comment (justinswart) [inline Library/Styles/CommentCellStyles.swift:42]: Let's keep this value as sentence-case ("Superbacker") and convert it to uppercase with a modifier.

File: Library/ViewModels/CommentCellViewModel.swift
Comment (justinswart) [inline Library/ViewModels/CommentCellViewModel.swift:12]: Let's move this to the cell and test with snapshots as per other comment.

File: Kickstarter-iOS/Views/Cells/CommentCell.swift
Comment (justinswart) [inline Kickstarter-iOS/Views/Cells/CommentCell.swift:65]: Can we rename the instances where we've referred to the `user` as `viewer` to `user`, we don't have the concept of a _viewer_ in the codebase.

File: Kickstarter-iOS/Views/Cells/CommentPostFailedCell.swift
Comment (justinswart) [inline Kickstarter-iOS/Views/Cells/CommentPostFailedCell.swift:91]: Need to confirm with copy that there is no full-stop at the end of this sentence 🤔 I see it does not appear in the designs.

File: Library/Styles/CommentCellStyles.swift
Comment (singhhari) [reply Library/Styles/CommentCellStyles.swift:53]: @justinswart can we reuse an existing "You" string or should we also include this in the next set of translations?

Comment (justinswart) [reply Library/Styles/CommentCellStyles.swift:53]: Yeah should be fine to reuse that string I think.

File: Kickstarter-iOS/Views/Cells/CommentPostFailedCell.swift
Comment (moderateepheezy) [reply Kickstarter-iOS/Views/Cells/CommentPostFailedCell.swift:91]: Paulo said we can use it for now https://kickstarter.slack.com/archives/CG55BBBBL/p1621979337078500

**Review by justinswart — APPROVED** [review]
Good to merge once CI passes! 👍

**Pull request merged**
